### ISSUE TITLE
Extending object syntax of v-on to support modifiers

### ISSUE THREAD

**Issue by brianwhu** [vuejs/vue#7846]
### What problem does this feature solve?
The object syntax of the v-on directive is vital in achieving conditional binding of event listeners, as shown in the discussions at https://github.com/vuejs/vue/issues/7349.

    <div v-on="{ mouseover: condition ? handler : null }">

However, the current object syntax does not allow modifiers.

This feature request suggest that we extend the object syntax in the following way to allow modifiers.

    <div v-on="{ click: { left: { prevent: condition ? leftClickHandler : null } },  mouseover: { stop: mouseoverHandler } }">

The above example would conditionally install leftClickHandler on "click.left.prevent" and mouseoverHandler on "mouseover.stop".

The embedded object notation is also conceptually consistent with the dot-notation already adopted in both function and inline syntax.

### What does the proposed API look like?
The proposed v-on object syntax would like like this, which is an extension of the current syntax.

    <div v-on="{ click: { left: { prevent: condition ? leftClickHandler : null } },  mouseover: { stop: mouseoverHandler } }">

**Comment by andrewsm80** [comment]
I think the below syntax would be better (no object embedding):

```vue
<div v-on="{ 'click.left.prevent': condition ? leftClickHandler : null,  'mouseover.stop': mouseoverHandler }">
```

**Comment by posva** [comment]
This is possible today by using https://vuejs.org/v2/guide/render-function.html#Event-amp-Key-Modifiers

```vue
<button v-on="{ '~click': () => foo = new Date() }">Trigger only once</button>
```

**Comment by Zsavajji** [comment]
Hi! I'm sorry to bring up such an old issue, but with the update to 2.6.11 i found that a thing related to this i used a lot in my project broke down.
I'm attaching `v-on:click` and `v-on:click.native` to a `<component` tag, which can dynamically be a Vue component or a DOM element. 
This now emits warnings, flooding logs and affecting development performance. 
I tried to attach those listeners dynamically but there's no object syntax for the `.native` modifier. How can i handle this?

**Comment by posva** [comment]
@Zsavajji if you think you found a bug please file a new issue with a  boiled down reproduction

**Comment by Zsavajji** [comment]
Not sure if it is a bug or expected behavior. Will open a bug report tomorrow :)

**Comment by magistr4815** [comment]
Hello.
I tried to do it like this, but it does not work.
```js
// not working
v-on="{ [condition ? 'click.stop' : 'click'] : eventfunc }"
// or
// error
v-on="{ condition ? 'click.stop' : 'click' : eventfunc }"
```
You need to create a variable.

```js
// in created
this.blabla = condition ? 'click.stop' : 'click';

// in template
v-on="{ blabla : eventfunc }"
```
Is there any other solution?
Thanks in advance!

**Comment by madmod** [comment]
This does not seem to work for all event methods. Only those which have the proprietary prefixes.

**Comment by Glandos** [comment]
> This is possible today by using [vuejs.org/v2/guide/render-function.html#Event-amp-Key-Modifiers](https://vuejs.org/v2/guide/render-function.html#Event-amp-Key-Modifiers)
> 
> ```
> <button v-on="{ '~click': () => foo = new Date() }">Trigger only once</button>
> ```

This seems nice, but undocumented. However, the `.native` modifier has no prefix, and seems unsupported with the object syntax. Is there any workaround for now?

**Comment by MarcelloTheArcane** [comment]
There doesn't seem to be an event amp key modifier for `.native`, which is what I need.

**Comment by Tyler-RCSD** [comment]
I need `.prevent`

**Comment by 3nuc** [comment]
[This is now documented here](https://v3.vuejs.org/api/directives.html#v-on):
>  v-on also supports binding to an object of event/listener pairs without an argument. **Note when using the object syntax, it does not support any modifiers.**

This means `{ 'keyup.left': onLeftRelease }` doesn't work. The alternative is obviously to handle this the old school way - [the native JS equivalents of some modifiers are described here](https://v3.vuejs.org/guide/render-function.html#v-on) (Event Modifiers section)

On one hand I know v-on object syntax is rarely used, however it's surprising this doesn't exist. As rationale, Vue is doing something similar with [multiple v-models](https://v3.vuejs.org/guide/component-custom-events.html#multiple-v-model-bindings) no? Seeing the magic behavior of emitting `update:propName` sort of has the same vibe as passing the magic string `keyup.left` to v-on. This probably reads funny but hopefully you can see where I'm coming from.

**Comment by sarayourfriend** [comment]
I previously ran into this issue (you can see the PR where I linked here in the feed above about a month ago) and then today I ran into this same problem again. Not content with not having a solution, I kept digging around, and discovered that you can use [`withModifiers`](https://vuejs.org/api/render-function.html#withmodifiers) (you can import from `@vue/runtime-dom`, though it is marked as `@private` :shrug:). You can see it working on this codesandbox:

https://codesandbox.io/s/nice-brown-j4q94s?file=/src/components/HelloWorld.vue:2364-2565

In fact, I randomly tried to get it working with multiple different listeners (because I wanted to be able to bind both a capture and non-capture event listener) and it totally works.

I have no idea if this is documented anywhere. I'd searched the Vue documentation so carefully before for _some_ solution to this problem. I have also not looked into how this is working in the Vue source code to tell whether this is working by accident or by design (or if there are any significant drawbacks to using this method.

Hope this helps someone else too :grinning:

**Comment by xia-2** [comment]
This problem still at vue3. 

`v-on supports binding to an object but it does not support any modifiers.`

**Comment by janverhulst** [comment]
I ended up creating a wrapper component like this:

```
<template>
  <router-link v-bind="$attrs" v-on="$listeners" @click.native="onClick">
    <slot />
  </router-link>
</template>

<script>
export default {
  name: 'RouterLinkWithClickEvent',
  methods: {
    onClick(...args) {
      this.$emit('click', ...args);
    }
  }
};
</script>
```

**Comment by Tenrys** [comment]
For anyone coming here from Google, there seems to be an builtin way to do this in Vue 3 now: https://vuejs.org/api/render-function.html#withmodifiers

**Comment by AvWoN** [comment]
For anyone happening upon this (btw, this should be fully supported by Vue via "click.capture" or camelCase syntax but alas no progress on this issue so far) the following works as well. Its a combination of the [`withModifiers`  & `v-on`](https://vuejs.org/guide/extras/render-function.html#v-on) with named event camelCase modifiers:
```
<script setup lang="ts">
// Vue v3.5.13

import {EventHandlers} from "~/components/player/types/types";

const onBox: EventHandlers = {
    "pointerdown": (e) => {
        console.log("Box received event", e);
        alert("Received!", e);
    },
}

const onParent: EventHandlers = {
    "pointerdownCapture": withModifiers((e) => {
        clickingParent(e)
    }, [ 'stop' ]),
}

// OR

const onParent2: EventHandlers = {
    "onPointerdownCapture": withModifiers((e) => {
        clickingParent(e)
    }, [ 'stop' ]),
}

const clickingParent = (e: MouseEvent) => {
    console.log("Parent received event", e);
}

</script>

<template>
    <div class="w-40 h-40 outline" v-on="onParent">
        <div class="w-10 h-10 bg-amber-200" v-on="onBox" @click="() => console.log('Box clicked')"/>
        Option 1
    </div>
    <div class="w-40 h-40 outline" v-bind="onParent2">
        <div class="w-10 h-10 bg-amber-200" v-on="onBox" @click="() => console.log('Box clicked')"/>
        Option 2
    </div>
</template>
```
It is scuffed since `Once` does not work but `Capture` does for whatever reason, even though all options are specified in the [v-on doc](https://vuejs.org/guide/extras/render-function.html#v-on) (`once` option probably isn't used by a single modern dev anyhow)
